crates/hotshot/src/traits/networking/push_cdn_network/definition/message_hook.rs

```diff
@@ -0,0 +1,536 @@
+pub struct HotShotMessageHook {
+    /// The cache for message hashes. We use this to deduplicate a sliding window of
+    /// 100 messages.
+    message_hash_cache: LruCache<u64, ()>,
+
+    /// The sample check interval
+    sample_check_interval: Duration,
+
+    /// The sample commit interval
+    sample_commit_interval: Duration,
+
+    /// The multiple of our average that the local average is allowed to be
+    allowed_multiple: u64,
+
+    /// The global moving average for the number of broadcast bytes per second
+    global_broadcast_bps: Sma,
+
+    /// The local average for the number of broadcast bytes per second
+    local_broadcast_bps: Sample,
+
+    /// The global moving average for the number of direct bytes per second
+    global_direct_bps: Sma,
+
+    /// The local average for the number of direct bytes per second
+    local_direct_bps: Sample,
+}
+
+impl Default for HotShotMessageHook {
+    /// # Panics
+    /// If 100 < 0
+    fn default() -> Self {
+        Self {
+            sample_check_interval: Duration::from_secs(5),
+            sample_commit_interval: Duration::from_secs(120),
+            allowed_multiple: 4,
+
+            global_broadcast_bps: Sma::new(),
+            global_direct_bps: Sma::new(),
+            local_broadcast_bps: Sample::new(),
+            local_direct_bps: Sample::new(),
+            message_hash_cache: LruCache::new(NonZeroUsize::new(100).unwrap()),
+        }
+    }
+}
+
+impl HotShotMessageHook {
+    /// Create a new `HotShotMessageHook`
+    ///
+    /// # Panics
+    /// If 100 < 0
+    #[must_use]
+    pub fn new(
+        sample_check_interval: Duration,
+        sample_commit_interval: Duration,
+        allowed_multiple: u64,
+    ) -> Self {
+        Self {
+            sample_check_interval,
+            sample_commit_interval,
+            allowed_multiple,
+
+            global_broadcast_bps: Sma::new(),
+            global_direct_bps: Sma::new(),
+            local_broadcast_bps: Sample::new(),
+            local_direct_bps: Sample::new(),
+
+            message_hash_cache: LruCache::new(NonZeroUsize::new(100).unwrap()),
+        }
+    }
+
+    /// Process a message against the moving average
+    /// Returns whether or not the message should be skipped.
+    fn process_against_sma(&mut self, message_len: usize, message_type: MessageType) -> HookResult {
+        // Match the sample and `SMA` based on the message type
+        let (sample, sma) = match message_type {
+            MessageType::Broadcast => (
+                &mut self.local_broadcast_bps,
+                &mut self.global_broadcast_bps,
+            ),
+            MessageType::Direct => (&mut self.local_direct_bps, &mut self.global_direct_bps),
+        };
+
+        // Get the current time
+        let now = Instant::now();
+
+        // Commit the sample if that interval has elapsed
+        if now.duration_since(sample.last_committed_time) >= self.sample_commit_interval {
```

Comment:
even if we are in cool we still add to global sma? if we are in cooldown for a while wont bytes_sent for sample always be 0?

Reply:
also seems below we have this logic probably want to remove it as we reset sample and second one shouldnt get hit

Reply:
My reasoning for adding to the global while still in cooldown was that it should count for _something_ since the CDN still is burdened with receiving it. But now that you mention it, it does probably make more sense to increment it after

Reply:
For adding to the global during cooldown: I wanted to avoid the chance that everyone is in cooldown and nobody is committing anything. Even if you send a really big message once and get put in jail I still think it should count

Reply:
Good catch on the duplicate logic, will remove it